### DISCUSSION TITLE
Subquery in column_property query issue

### DISCUSSION THREAD

**Discussion by aaronshifman** [sqlalchemy/sqlalchemy#7085] [Usage Questions] [answered]
I have a table that looks something like this

| id | class_type | date |
|----|------------|------|
| 0  | 0          | 2020 |
| 1  | 0          | 2021 |
| 2  | 1          | 2022 |

And I'd like to determine the "version" within the `class_type` such that I have a Mapper that looks like

| id | class_type | date | version |
|----|------------|------|---------|
| 0  | 0          | 2020 | 1       |
| 1  | 0          | 2021 | 2       |
| 2  | 1          | 2022 | 1       |

My current solution is as follows is in my Table class I have

```python
    @classmethod
    def __declare_last__(cls):
        sub_q = select(cls.id.label("id"),
                       func.row_number().over(partition_by=[cls.class_type],
                                              order_by=cls.date.asc()).label('row_number'))

        cls.version = column_property(
            select(sub_q.c.row_number).where(sub_q.c.id == cls.id)
        )
```

This works if I say `Session.query(Table.id, Table.version).all()`

However `Session.query(Table.version)all()`

Crashes and complains about "(1242, 'Subquery returns more than 1 row')" 

The SQL emitted is 

```sql
SELECT (SELECT anon_2.`row_number` 
FROM (SELECT table.id AS id, row_number() OVER (PARTITION BY table.class_type ORDER BY table.date ASC) AS `row_number` 
FROM table) AS anon_2, table 
WHERE anon_2.id = table.id) AS anon_1
```
And it looks like there's a redundant select 

I'm sure I'm doing something stupid but I can't quite figure it out

**Comment by zzzeek** [comment]
im not even sure what you have there was ever tested, so this could be a bug.  can you provide your models (truncated, jut the columns needed to produce the query) ?  I can paste that query and see what it's doing.  normally the issue is that you need to apply correlate() to the subquery correctly but this looks a little more weird than that.

**Reply by aaronshifman** [reply]
Thanks for the quick reply. Sure thing - let me grab them

**Comment by aaronshifman** [comment]
To give context to this. The goal is to have `version` be relative to the entire table and not to the query (however small the subset may be). I.e. the ``` OVER PARTITION``` part is a subquery. If there's a better way to accomplish that it might be easier to do that.

**Comment by aaronshifman** [comment]
I believe this should be it

```python
"""Specify and create the database structure and relationship mappings."""

from sqlalchemy import Column, Integer, text, func
from sqlalchemy import select
from sqlalchemy.dialects.mysql import DATETIME
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import column_property

Base = declarative_base()


class Table(Base):
    __tablename__ = 'table'

    id = Column(Integer, primary_key=True, autoincrement=True)
    class_type = Column(Integer)

    date_registered = Column(DATETIME(fsp=6), nullable=False, server_default=text('CURRENT_TIMESTAMP(6)'))

    @classmethod
    def __declare_last__(cls):
        sub_q = select(cls.id.label("id"),
                       func.row_number().over(partition_by=[cls.class_type],
                                              order_by=cls.date.asc()).label('row_number'))

        cls.version = column_property(
            select(sub_q.c.row_number).where(sub_q.c.id == cls.id)
        )

```

**Comment by zzzeek** [comment, marked as answer]
oh OK, I just came up with this quick:

```py
from sqlalchemy import *
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import *
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String)
    class_type = Column(String)
    date = Column(DateTime)

    @classmethod
    def __declare_last__(cls):
        sub_q = select(cls.id.label("id"),
                       func.row_number().over(partition_by=[cls.class_type],
                                              order_by=cls.date.asc()).label('row_number'))

        cls.version = column_property(
            select(sub_q.c.row_number).where(sub_q.c.id == cls.id)
        )

configure_mappers()

s = Session()

print(s.query(A.version))
```

first thing is use subquery() and scalar_subquery(), there are two warnings generated that you should see.  so like this:

```py
class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)
    class_type = Column(String)
    date = Column(DateTime)

    @classmethod
    def __declare_last__(cls):
        sub_q = select(
            cls.id.label("id"),
            func.row_number()
            .over(partition_by=[cls.class_type], order_by=cls.date.asc())
            .label("row_number"),
        ).subquery()

        cls.version = column_property(
            select(sub_q.c.row_number)
            .where(sub_q.c.id == cls.id)
            .scalar_subquery()
        )

```

that doesnt fix the problem here though, because the ORM seems to not know what its selecting from, so immediately it can work like this:
```
   print(s.query(A.version).select_from(A))
```
gives us

```
SELECT (SELECT anon_2.row_number 
FROM (SELECT a.id AS id, row_number() OVER (PARTITION BY a.class_type ORDER BY a.date ASC) AS row_number 
FROM a) AS anon_2 
WHERE anon_2.id = a.id) AS anon_1 
FROM a
```

that's as far as I can get it for now.  need to give it a FROM clause

**Comment by aaronshifman** [comment]
That seems to work exactly as I would need it - thank you @zzzeek . I can tack on the .select_from where I pull `version` on its own.  Since apparently this is a bit of an oddball approach - is there a better way to do  it?

**Reply by zzzeek** [reply]
probably not right now.  these column_property() with subuqery things are a PITA

**Reply by aaronshifman** [reply]
Fair - thanks for all the help
